### DISCUSSION TITLE
PC/GEOS sound very low / quiet in comparison to running under Dosbox

### DISCUSSION THREAD

**Discussion by bolle732** [dosemu2/dosemu2#2280] [Q&A]
Running on Ubuntu 24.04. I've been playing with FM sound under PC/GEOS (FreeGEOS). It has a Soundblaster 1.0 / 1.5 driver. When playing FM sounds, they are very low / quite under Dosemu2. Under Dosbox, it's like 3x louder.

Ubtuntu is using Pipewire. In the mixer I see Dosemu2 as "libao[dosemu.bin]". The Dosbox is listed listed directly "dosbox". Both Dosemu2 and Dosbox are pristine regarding sound.

Any idea, how I can make FM sound louder under Dosemu2?

PS:
Sorry for comparing Dosemu2 with Dosbox...

**Comment by stsp** [comment]
https://github.com/dosemu2/install-freedos/issues/13
At the bottom of the thread
you'll find SBPMIXER.zip.
I am not sure if it is already a
part of freedos or not, but you
can use it.

**Reply by bolle732** [reply]
Thanks for the tip. I gave them a try. Unfortunately, it isn't any louder even with `SBPMIXER /M 15 /V 15 /F 15`. This values defaulted to `12`.

**Comment by stsp** [comment]
> Sorry for comparing Dosemu2 with Dosbox...

I don't think this is something to
be sorry about. Most people that
try dosemu2, have dosbox too.
And while dosbox was always a
piece of junk (never worked for
me properly), I have to admit that
dosbox-staging actually works
most of the time.

**Reply by bolle732** [reply]
Yeah, maybe I'll have to cleanup old memories which are not true anymore...

**Comment by bolle732** [comment]
By using Alsamixer, I can upper the volume of Dosemu2. But then everything other then Dosemu2 get very loud.

**Comment by bolle732** [comment]
So, the base volume of Dosemu2 is lower then any other application.

**Comment by stsp** [comment]
With something like this:
```
diff --git a/src/base/dev/sb16/dspio.c b/src/base/dev/sb16/dspio.c
index 5c62fecd0..7120cd49b 100644
--- a/src/base/dev/sb16/dspio.c
+++ b/src/base/dev/sb16/dspio.c
@@ -897,7 +897,7 @@ static double dspio_get_volume(int id, int chan_dst, int chan_src, void *arg)
 double dspio_calc_vol(int val, int step, int init_db)
 {
 #define LOG_SCALE 0.02
-    return pow(10, LOG_SCALE * (val * step + init_db));
+    return pow(10, LOG_SCALE * (val * step * 2 + init_db));
 }
 
 static int dspio_is_connected(int id, void *arg)
```
you can fell from a chair. :)

**Reply by bolle732** [reply]
All right, I'm back on the chair ;-)

Really interesting is, that after playing with Alsamixer, the sound is in the expected range.

So, at the beginning, the Gnome settings volume control has been at around 75% and in the Alsamixer, I saw it first at around 40%. At this moment, Dosemu2 was too quiet. After playing with the master volume in Alsamixer, the Gnome volume changes accordingly with the master volume of Alsamixer. And now, Dosemu2 is much louder then before when compared to the volume configured. Even after terminating and executing Dosemu2 again.

I'll have to restart this machine and recheck again. That would be at the end of the week because this machine runs 7/24. But I have a second machine to test. Let's see what I get there.

At least Dosbox is too loud for my taste. Ok, you can control the volume per application. But I think, that the default volumes of i.e. Youtube in browser, a music player or voice chats are in similar range. Currently Dosbox is over and Dosemu2 is below it really noticeable. And I don't know if this is something that depends on the distribution / sound stack and/or the applications.

**Comment by stsp** [comment]
One thing you can check is if
that dspio_calc_volume() returns
values around 1.0 when all volumes
set on maximum. Also sbpmixer
doesn't expose Gain control that
can go above 1.0.
So if dspio_calc_volume() is correct,
then there is not too much I can do.
Youtube player doesn't normalize
the volume for you, too, but there is
a convention among youtubers. Still
I constantly hit videos that are too
loud or too soft. Volume slider comes
at a rescue, but there isn't any in
dosemu. IMHO the app should be
able to request the volume slider
among the window decorations.
Something to suggest to freedesktop
people. :)

Вторник, 1 октября 2024 г получено от bolle732:

**Comment by bolle732** [comment]
I've tested now on my second machine. It behaves the same. No wonder, both machines have Ubuntu 24.04 and the same installation. Dosemu2 is on both a lot quieter then Dosbox, even compared to any other Linux programs.

Using a factor of 2 in `dspio_calc_vol` makes it sounding too loud. Everything gets distorted so much, that even my (*) metal accustomed brain get blown away ;-) But using a factor of 1.4 get it close to that of Dosbox without overdrive. Haven't checked the result for going close to 1.0.

Regarding a volume slider, I'll would prefer a CLI tool or internal command to control this. That can be used like `emumouse` and makes one able to use it in a batch for different purposes.

One thing that would be nice is having a list of reference tools to be used together with Dosemu2 for testing and comparing. Not only for sound, also for video, network and more. Anyone should check first with this tools and then come here for discussing his problems providing the results of the reference tools.

**Reply by stsp** [reply]
> Dosemu2 is on both a lot quieter then Dosbox, even compared to any other Linux programs.

Is this with default, or after raising
things in sbpmixer? For me changing
Master from 12 to 15 changes a lot.

> Regarding a volume slider, I'll would prefer a CLI tool or internal command to control this.

No, this isn't something mutually exclusive.
Modern desktops already support the
concept of per-app volume. You can run
`pavucontrol` and increase the volume
exactly for dosemu2. Such a volume
control should be possible to request
for the window decorators, but it will
still be handled by PulseAudio, not by
the app itself.

> That can be used like emumouse and makes one able to use it in a batch for different purposes.

Please ask Shidel to add the Gain
control to sbpmixer. You can also
try sb16mix from Creative, you
only need to set `SOUND=C:\SB16`
for it to work. It has Gain control.

What reference tools? The ones
to measure the sound volume
coming out of your speakers?

**Reply by stsp** [reply]
> Using a factor of 2 in dspio_calc_vol makes it sounding too loud.

I think it makes sense to multiple
the return value of the function by
2, rather than the exponential part.
Multiplying exponent part was a
joke to get you out of a chair.

**Comment by bolle732** [comment]
> Is this with default, or after raising
> things in sbpmixer? For me changing
> Master from 12 to 15 changes a lot.

This was with default 12.

> No, this isn't something mutually exclusive.
> Modern desktops already support the
> concept of per-app volume. You can run
> pavucontrol and increase the volume
> exactly for dosemu2. Such a volume
> control should be possible to request
> for the window decorators, but it will
> still be handled by PulseAudio, not by
> the app itself.

Yes, that's true that we can control the loudness by application.

> Please ask Shidel to add the Gain
> control to sbpmixer. You can also
> try sb16mix from Creative, you
> only need to set SOUND=C:\SB16
> for it to work. It has Gain control.

Ok, I'll give this a try. But still would love it to  be done directly from Doseum2 without any external tool.

> What reference tools? The ones
> to measure the sound volume
> coming out of your speakers?

No, they should only generate a sound at a specific loudness. May sinus wave and/or white / rose noise. With them and standard settings in Dosemu2 you have a reference.

> I think it makes sense to multiple
> the return value of the function by
> 2, rather than the exponential part.
> Multiplying exponent part was a
> joke to get you out of a chair.

All right ;-) Multiplying the whole return value with 2 gives some snap noises. But with 1.9 it's ok.

**Reply by stsp** [reply]
> This was with default 12.

Then use `userhook.bat` to adjust.
Does this solve the problem?
Or gain is also needed?

> Yes, that's true that we can control the loudness by application.

Yes, and its saved permanently.
Why the app still can't request the
volume slider, is not quite clear.

> But still would love it to be done directly from Doseum2 without any external tool.

This is why I'd like it to be able to
display the system's per-app control.
Then you can do that "from dosemu2"
but no new code in dosemu2 is needed.
I am not sure if SDL provides an API
to the system volume control - I think
it only has its own software volume
adjustments.

> Under Dosemu2 it starts stuttering

You mean midi?
Does it say something in terminal?

**Reply by bolle732** [reply]
> > This was with default 12.
> 
> Then use `userhook.bat` to adjust. Does this solve the problem? Or gain is also needed?

Increasing to 15 helps a little, but not enough in my "ears". I'll have to check what the gain will do.
 
> > But still would love it to be done directly from Doseum2 without any external tool.
> 
> This is why I'd like it to be able to display the system's per-app control. Then you can do that "from dosemu2" but no new code in dosemu2 is needed. I am not sure if SDL provides an API to the system volume control - I think it only has its own software volume adjustments.

Maybe SDL will hide this because it's something that not all platforms it runs on supports? 

> > Under Dosemu2 it starts stuttering
> 
> You mean midi? Does it say something in terminal?

As far as I understand, this sound is rendered and played as FM. Not MIDI (which PC/GEOS unfortunately doesn't support).

**Reply by stsp** [reply]
> Maybe SDL will hide this because it's something that not all platforms it runs on supports?

Which is why I think it should be
done on the window decoration
level. In this case if its not supported
by desktop, then there is no control.

> As far as I understand, this sound is rendered and played as FM.

So how is it stuttering, with a small
periods of silence, or just some note
sticks for longer than it should?
Any errors on terminal?

**Reply by bolle732** [reply]
> So how is it stuttering, with a small
> periods of silence, or just some note
> sticks for longer than it should?
> Any errors on terminal?

It's playing all the notes, but the times between the notes is sometimes too short or too long. Like going faster and slower or stumbling. It's FM, so I believe the question is from where comes the time source and what makes that source be unstable / non-monotone.

**Reply by bolle732** [reply]
Regarding error, I'll have to check by starting Dosemu2 from terminal.

**Reply by bolle732** [reply]
Currently no warnings on console. I remember to have seen some PCM warnings regarding time earlier. But for now nothing.

**Comment by bolle732** [comment]
Question between:

Is there a hint regarding stuttering sound? With Dosbox, my test sounds plays all the time fine. Under Dosemu2 it starts stuttering from time to time. Is that an effect by the CPU emulation?

**Reply by stsp** [reply]
Today I reverted the patch that did
stuttering in StreetFighter2 game.
So you can try an update, although
I don't think it's gonna be the same
problem.

**Reply by stsp** [reply]
And yesterday I reverted another
patch that did stuttering in Wolf3d...
Actually updating is worth a try.

**Reply by bolle732** [reply]
Yes, I'm a little behind `devel` as I'm working on a little app for PC/GEOS and wanted a non-changing environment to do the devs / tests. All that about the privileges sound really nice, but I postponed testing it with PC/GEOS. Even your work on IPX sound very interesting as this could maybe make the debugging tool of PC/GEOS (called Swat) to use Dosemu2 as the target. Currently, it uses Dosbox.

**Reply by stsp** [reply]
> Even your work on IPX sound very interesting as this could maybe make the debugging tool of PC/GEOS (called Swat)
> to use Dosemu2 as the target.

What does swat debugger have in
common with IPX/UDP relay?

**Reply by stsp** [reply]
> Yes, I'm a little behind devel as I'm working on a little app for PC/GEOS and wanted a non-changing
> environment to do the devs / tests.

I guess you just don't want to rebuild
the entire stack... Which is understandable,
but you could as well use PPA.

**Reply by bolle732** [reply]
> What does swat debugger have in
> common with IPX/UDP relay?

For Swat, there is always a target machine running the PC/GEOS. Either a dedicated physically computer connected by serial line or a "virtual" one connected by a anonymous pipe (Windows) or IPX connection.

In the Dosbox case, there is started an IPX server like this:
```
ipxnet startserver 8079
ipxnet status > c:\ipx_status.txt
```

Then on the host side, Swat is invoked like this (where IPX port comes from the `ipx_status.txt`):
swat -net 00000000:7F000001${IPX_PORT}:003F

So, there seems no UDP in use and therefore has nothing todo with IPX/UDP relay. When I saw your work regarding IPX, I was thinking that maybe there would be a way todo the same as with Dosbox.

**Reply by bolle732** [reply]
> I guess you just don't want to rebuild
> the entire stack... Which is understandable,
> but you could as well use PPA.

Yes, exactly. But with the PPA I'm unable to get blown from the chair ;-)

**Reply by stsp** [reply]
> In the Dosbox case, there is started an IPX server like this:

I don't think I understand.
`startserver 8079` AFAIK starts
UDP relay server on port 8079.
You can check it with `netstat`
tool, make sure its an UDP port.
Then it gets the virtual IPX node
address that only this server
is supposed to be aware about,
but you tell it to swat...
There are some missing bits of
info for sure.
Do you need the in-kernel IPX
support for this or not?

> Yes, exactly. But with the PPA I'm unable to get blown from the chair ;-)

You can install everything but
dosemu2 itself from PPA. You'd
need some `-dev` packages to
be installed (fdpp-dev, djdev64-dev
etc) to build it. `make deb` will
tell you what's missing.

**Reply by bolle732** [reply]
> I don't think I understand.
> startserver 8079 AFAIK starts
> UDP relay server on port 8079.
> You can check it with netstat
> tool, make sure its an UDP port.
> Then it gets the virtual IPX node
> address that only this server
> is supposed to be aware about,
> but you tell it to swat...
> There are some missing bits of
> info for sure.
> Do you need the in-kernel IPX
> support for this or not?

I really haven't enough insight into all this details and have to dig around more. Have to ask Falk from Blueways / FreeGEOS. But at least, this would be something to investigate more? At least if it's not a dead end on the foresight.

**Reply by bolle732** [reply]
> You can install everything but
> dosemu2 itself from PPA. You'd
> need some -dev packages to
> be installed (fdpp-dev, djdev64-dev
> etc) to build it. make deb will
> tell you what's missing.

Building Dosemu2 from source works fine. At least when I did like 2 months ago. You mean installing PPA version at parallel?

**Reply by stsp** [reply]
I mean, install fdpp, comcom64,
dj64dev and all that from PPA, so
that when you build dosemu2, you
don't build 1024+ deps.

**Reply by bolle732** [reply]
Yeah, but then I loose the ability to hack around the in the code? I'll need that from time to time.. Or I should start using VMs...

**Reply by stsp** [reply]
Why if you build dosemu2 itself
from source? Are you hacking fdpp,
comcom64 or dj64dev? I don't
think so.

**Reply by bolle732** [reply]
> Why if you build dosemu2 itself
> from source? Are you hacking fdpp,
> comcom64 or dj64dev? I don't
> think so.

Sometimes I digged into fdpp and comcomNN when I had problems with running DOS programs. But yes, it's very a rare case. So, give me some time to get up to current devel.
